pkg/resources/server.go

```diff
@@ -118,18 +122,46 @@ func (rs *resourceServer) NotifyRegistrationStatus(ctx context.Context,
 func (rs *resourceServer) Allocate(ctx context.Context, rqt *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
 	glog.Infof("Allocate() called with %+v", rqt)
 	resp := new(pluginapi.AllocateResponse)
+
+	err := rs.cdiCache.Refresh()
+	if err != nil {
+		glog.Errorf("failed to refresh CDI spec: %v", err)
+		return nil, err
+	}
 	for _, container := range rqt.ContainerRequests {
 		containerResp := new(pluginapi.ContainerAllocateResponse)
-		containerResp.Devices = rs.resourcePool.GetDeviceSpecs(container.DevicesIDs)
 
 		envs, err := rs.getEnvs(container.DevicesIDs)
 		if err != nil {
 			glog.Errorf("failed to get environment variables for device IDs %v: %v", container.DevicesIDs, err)
 			return nil, err
 		}
-		containerResp.Envs = envs
 
-		containerResp.Mounts = rs.resourcePool.GetMounts(container.DevicesIDs)
+		if rs.useCdi {
+			containerResp.Annotations = make(map[string]string, 0)
+
+			for _, id := range container.DevicesIDs {
```

Comment:
device IDs to container annotation can be a function/method in cdi pkg (related to previous comment)